File: frontend-design/scripts/lookup_term.py
#!/usr/bin/env python3
"""Lookup canonical frontend-design terminology.

Usage:
  python scripts/lookup_term.py neubrutalism
  python scripts/lookup_term.py "soft ui"
"""

from __future__ import annotations

import argparse
from _catalog import load_csv, rank_csv, pick, md_bullet


def main() -> int:
    parser = argparse.ArgumentParser(description="Lookup canonical design terminology")
    parser.add_argument("query", nargs="+")
    parser.add_argument("-n", "--limit", type=int, default=5)
    args = parser.parse_args()
    query = " ".join(args.query)
    query_l = query.lower()
    all_rows = load_csv("terminology.csv")
    exact = []
    for row in all_rows:
        names = [pick(row, "canonical"), *[item.strip() for item in pick(row, "aliases").split(";") if item.strip()]]
        if any(name.lower() == query_l for name in names):
            exact.append(row)
    rows = exact or rank_csv("terminology.csv", query, args.limit)
    if not rows:
        print("No matching term")
        return 1
    print(f"# Terminology Lookup: {query}\n")
    for row in rows:
        print(f"## {pick(row, 'canonical')}")
        print(md_bullet("Kind", pick(row, "kind")))
        print(md_bullet("Aliases", pick(row, "aliases")))
        print(md_bullet("Definition", pick(row, "definition")))
        print(md_bullet("Notes", pick(row, "notes")))
        print("")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend-design/scripts/lookup_term.mjs
#!/usr/bin/env node
/** Lookup canonical frontend-design terminology.
 *
 * Usage:
 *   node scripts/lookup_term.mjs neubrutalism
 *   node scripts/lookup_term.mjs "soft ui"
 */

import { pathToFileURL } from "node:url";
import { loadCsv, mdBullet, pick, rankCsv } from "./_catalog.mjs";

function parseArgs(args) {
  let limit = 5;
  const query = [];
  for (let index = 0; index < args.length; index += 1) {
    if (args[index] === "-n" || args[index] === "--limit") {
      limit = Number.parseInt(args[index + 1], 10);
      index += 1;
    } else {
      query.push(args[index]);
    }
  }
  if (query.length === 0 || !Number.isInteger(limit)) throw new Error("Usage: lookup_term.mjs <query> [-n <limit>]");
  return { query: query.join(" "), limit };
}

export function main(args = process.argv.slice(2)) {
  let options;
  try {
    options = parseArgs(args);
  } catch (error) {
    console.error(error.message);
    return 2;
  }
  const normalized = options.query.toLowerCase();
  const allRows = loadCsv("terminology.csv");
  const exact = allRows.filter((row) => {
    const names = [pick(row, "canonical"), ...pick(row, "aliases").split(";").map((item) => item.trim()).filter(Boolean)];
    return names.some((name) => name.toLowerCase() === normalized);
  });
  const rows = exact.length > 0 ? exact : rankCsv("terminology.csv", options.query, options.limit);
  if (rows.length === 0) {
    console.log("No matching term");
    return 1;
  }
  const lines = [`# Terminology Lookup: ${options.query}`, ""];
  for (const row of rows) {
    lines.push(
      `## ${pick(row, "canonical")}`,
      mdBullet("Kind", pick(row, "kind")),
      mdBullet("Aliases", pick(row, "aliases")),
      mdBullet("Definition", pick(row, "definition")),
      mdBullet("Notes", pick(row, "notes")),
      "",
    );
  }
  process.stdout.write(`${lines.join("\n")}\n`);
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) process.exitCode = main();
